docs/registry/README.md

```diff
@@ -0,0 +1,193 @@
+# Registry Proposal
+
+## Objective
+
+Design an **initial** version of the Registry that can support discoverability of 
+the different event types that can be consumed from the eventing mesh.
+
+## Requirements
+
+Our design revolves around the following core requirements:
+
+1. We should have a Registry per namespace to enforce isolation.
+2. The Registry should contain the event types that can be consumed from
+the eventing mesh. If an event type is not ready for consumption, we should explicitly indicate so.
+3. The event types stored in the Registry should contain all the required information 
+for a consumer to create a Trigger without resorting to some other OOB mechanism.
+
+## Design Ideas
+
+### EventType CRD
+
+We propose having a namespaced EventType CRD. Here is an example of how a CR would look like:
+
+```yaml
+apiVersion: eventing.knative.dev/v1alpha1
+kind: EventType
+metadata:
+  name: repopush
+spec:
+  type: repo:push
+  source: my-user/my-repo
+  schema: my-schema
+  broker: default
+```
+
+
+- The `name` of the EventType is advisory, non-authoritative. Given that Cloud Event types can 
+contain characters that may not comply with Kubernetes naming conventions, we will (slightly) 
+modify those names to make them K8s-compliant, whenever we need to generate them. 
+
+- `type` is authoritative. This refers to the Cloud Event type as it enters into the eventing mesh. 
+
+- `source`: an identifier of where we receive the event from. This might not necessarily be the Cloud Event source 
+attribute. If we receive the event from our receive adaptors, the info might come in a Cloud Event custom extension (e.g., from).
+
+- `schema` is a URI with the EventType schema. It may be a JSON schema, a protobuf schema, etc. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. 
+
+### EventType Instantiation
+
+We foresee the following ways of populating the Registry:
+
+**1. Event Source CR installation**
+
+Upon installation of an Event Source CR, the source will register its EventTypes.
+
+Example:
+
+```yaml
+apiVersion: sources.eventing.knative.dev/v1alpha1
+kind: GitHubSource
+metadata:
+  name: github-source-sample
+  namespace: default
+spec:
+  eventTypes:
+    - push
+    - pull_request
+  ownerAndRepository: my-other-user/my-other-repo
+  accessToken:
+    secretKeyRef:
+      name: github-secret
+      key: accessToken
+  secretToken:
+    secretKeyRef:
+      name: github-secret
+      key: secretToken
+  sink:
+    apiVersion: eventing.knative.dev/v1alpha1
+    kind: Broker
+    name: default
+
+```
+ 
+By applying this file, two EventTypes will be registered, with types `push` and `pull_request`, 
+source `my-other-user/my-other-repo`, for the `default` Broker in the `default` namespace.
+
+**2. Manual User Registration**
```

Comment:
I guess it depends on who the "user" is, but I would think that most users (or even ops folks) wouldn't know what types to register. Rather, I would think that along with the yaml that defines the EventSource CRD would be  the yaml to create the eventTypes. I view there being a tight coupling between EventSources and the EventTypes - at least in the cases where the code in the EventSource twiddles the events and knows, in advance, about all of the event producer's events.  I'm not actually fond of eventSources knowing this info, but at least in the github case it appears to.

Reply:
Some event sources have the info, some don't. That is why we will also need the auto-registration of events upon first arrival, or maybe some way to query producers. The latter seems IMO more suitable for a second cut.

My first iteration of this was, as you said, having the EventTypes being created at the Event Source CRD-level. The problem with that is that you don't have a namespace at that point. Which namespace the EventType should be created on? I was basically having a controller for source CRDs, whenever one was created, I was then looking at all the namespaces with eventing enabled, and adding all the possible EventTypes from that CRD to each of those namespaces.
A big downside of this is that the registry information is *not accurate*, it only has info about EventTypes that may eventually flow in the system. But actually, they will only be able to flow if you create a CR.

Reply:
@nachocano @duglin Does auto-registration also include auto-deletion? I am really a fan of K8s extensibility with CRDs, but for event types I wonder, if this is the right mechanism.

Reply:
@deissnerk if we auto-register stuff, IMHO I think we should somehow support auto-deletion. For example when a k8s event source is removed, we should go ahead and remove all its EventTypes from the registry.

BTW, I created a separate [PR](https://github.com/knative/eventing/pull/1015) with just this README, not to pollute the design discussion with code. If you guys can start commenting on that one, that'd be great. I already made some updates to the README there.

Reply:
@nachocano Sorry, I just found the other PR. 😄

Reply:
@deissnerk no worries! Thanks for your comments and feedback!